### ISSUE TITLE
How to convert simple NER format to spacy json

### ISSUE THREAD

**Issue by r-wheeler** [explosion/spaCy#1966]
Currently I have data in the simple training data / offset format as shown in the docs:

```
TRAIN_DATA = [
    ('Who is Shaka Khan?', {
        'entities': [(7, 17, 'PERSON')]
    }),
    ('I like London and Berlin.', {
        'entities': [(7, 13, 'LOC'), (18, 24, 'LOC')]
    })
]
```

In order to utilize the CLI trainer, this data needs to be in spacys internal json training data format. to do this I am currently converting the simple offset format to conllner, then using the spacy cli converter. 

Is there a better way to directly get offset / simple format directly to spacy json format? Using the cli trainer seems to be able to train much faster than using the method specified [here](spacy/examples/training/train_ner.py)

## Your Environment
* **spaCy version:** 2.0.5
* **Platform:** Linux-4.4.0-1049-aws-x86_64-with-debian-stretch-sid
* **Python version:** 3.6.4
* **Models:** en_core_web_lg, en

**Comment by ines** [comment]
Yes, there's a [`gold.biluo_tags_from_offsets`](https://spacy.io/api/goldparse#biluo_tags_from_offsets) function that converts the entity offsets to a list of per-token BILUO tags:

```python
from spacy.gold import biluo_tags_from_offsets

doc = nlp(u'I like London.')
entities = [(7, 13, 'LOC')]
tags = biluo_tags_from_offsets(doc, entities)
assert tags == ['O', 'O', 'U-LOC', 'O']
```

**Comment by r-wheeler** [comment]
Is there a way to save the bilou tags in the json format used by cli
trainer?

**Comment by honnibal** [comment]
I think we currently lack a writer that does exactly that. I've been finding the current data-format situation frustrating too. Needs improvement.

**Comment by azarezade** [comment]
I have also the same problem in training NER.
@honnibal @r-wheeler 
Is there any json sample training data for NER?

**Comment by ines** [comment]
You can find an the structure and an example of the training JSON format here:
https://spacy.io/api/annotation#json-input

So you should be able to write a script that produces data in this format, using the BILUO tags. I agree that spaCy should make this easier (and, as @honnial mentioned above, we're also thinking about overhauling the training format in general to make it easier to work with).

**Comment by azarezade** [comment]
Thanks @ines 

That is a very good idea, I'm looking forward to it.

**Comment by rahulgithub** [comment]
Same question that I posted on the chat channel today. I want my program to point to training data in a separate file rather than having it in the code. I currently have as follows based on the examples on the website

<img width="1102" alt="screen shot 2018-04-03 at 2 44 06 pm" src="https://user-images.githubusercontent.com/4360712/38277517-838f5b46-374d-11e8-9b27-0b8331a3ad9b.png">

**Comment by DevaChandraRaju** [comment]
@rahulgithub, recently I was looking for similar need. I figured a way to put the content to a separate file instead of in code.
`inpt_text = open("text_file","r")`
`TRAIN_DATA = inpt_text.read().replace('\n ', '')`
`TRAIN_DATA = ast.literal_eval(' '.join(TRAIN_DATA.split()))`

**text_file content format:**
 [
    ('Who is Shaka Khan?', {
        'entities': [(7, 17, 'PERSON')]
    }),
    ('I like London and Berlin.', {
        'entities': [(7, 13, 'LOC'), (18, 24, 'LOC')]
    })
]

Plz share if there are any other approaches to maintain TRAIN_DATA in a separate file instead of in code.

**Comment by ines** [comment]
@rahulgithub @DevaChandraRaju How you solve this depends on which file formats you like to work with. `ast.literal_eval` isn't *bad* – but it's also a very uncommon way to pass data around. And the transformations (replacing, splitting, joining) you have to do to read in the file can very easily lead to bugs.

A simpler way might be working with JSON. Here's a an example of a file ```data.json```:

```json
[
    ["Who is Shaka Khan?", {"entities": [[7, 17, "PERSON"]]}],
    ["I like London and Berlin", {"entities": [[7, 13, "LOC"], [18, 24, "LOC"]]}]
]
```

JSON doesn't have tuples, so you need to use lists everywhere. In Python, you can then do the following:

```python
import json
import io

with io.open('/path/to/data.json', encoding='utf8') as f:
    train_data = json.load(f)
```

Also remember that you can always re-write the training script to format your data differently. If you're working with large training sets, we also recommend to use the `spacy train` command instead – this will take care of batching up your examples correctly for optimal performance, and it'll also give you much more useful command line output.

**Comment by SandeepNaidu** [comment]
I am also about to write a script to convert from inline training format to CLI training format. However I train the model on sentences. So each paragraph entry will contain exactly one sentence instead. Is that fine? Since the NER model trains on context which are usually no more than sentences, I do not see a reason for paragraphs to be input with their raw text and then sentences. I think paragraphs are there for Dependency Parsing. Is that right?

**Comment by SandeepNaidu** [comment]
I am pasting an answer derived from [stackoverflow](https://stackoverflow.com/questions/48916768/how-do-i-convert-simple-training-style-data-to-spacys-command-line-json-format) which is not showing up easily in searches. Ofcourse you will have to improvise a bit if you have multiple paragraphs.
```
def getcharoffsetsfromwordoffsets(doc,entities):
    charoffsets = []
    for entity in entities:
        span = doc[entity[0]:entity[1]]
        charoffsetentitytuple = (span.start_char, span.end_char, entity[2])
        charoffsets.append(charoffsetentitytuple)
    return charoffsets

def convertspacyapiformattocliformat(nlp, TRAIN_DATA):
    docnum = 1
    documents = []
    for t in TRAIN_DATA:
        doc = nlp(t[0])
        charoffsetstuple = getcharoffsetsfromwordoffsets(doc,t[1]['entities'])
        tags = biluo_tags_from_offsets(doc, charoffsetstuple)
        ner_info = list(zip(doc, tags))
        tokens = []
        sentences = []
        for n, i in enumerate(ner_info):
            token = {"head" : 0,
            "dep" : "",
            "tag" : "",
            "orth" : i[0].string,
            "ner" : i[1],
            "id" : n}
            tokens.append(token)
        sentences.append({'tokens' : tokens})
        document = {}
        document['id'] = docnum
        docnum+=1
        document['paragraphs'] = []
        paragraph = {'raw': doc.text,"sentences" : sentences}
        document['paragraphs'] = [paragraph]
        documents.append(document)
    return documents
```

**Comment by eswar3** [comment]
@ines is "spacy convert" command the solution for this issue? The reason why I am asking this is I don't see a converter for BILOU. Is that something currently mot supported?

**Comment by LizaKoz** [comment]
Hello everyone, 

I have the same question as @eswar3 about the input data format which I have to pass to convert function in order to train my NER for my custom entities. 

Right now, my data looks like simple NER format with about 10 labels, when I try to pass it to  the converter function it give me this error:   Unknown format Can't find converter for json. 

Could you please help me: what am I doing wrong?

**Comment by honnibal** [comment]
In v2.1.0a1, this is now finally be easier. Thanks for your patience with the annoying format, which has a few extra levels of nesting so that it can accomodate whole document annotations, along with paragraph structure.

In v2.1.0a1, there's now a new function `spacy.gold.docs_to_json`. This function takes a list of `Doc` objects and outputs properly formed json data, for use in `spacy train`. You just have to set up the `Doc` objects to have the annotations you want. For named entities, this is as easy as writing to the `doc.ents` attribute. For the parser, the best approach is to create an array with the head offsets you want, and then import the data with `doc.from_array()`.
